File: src/commands/doctor.ts
import type { CommandDef } from "citty";
import { defineCommand } from "citty";
import { consola } from "consola";
import { createCommandLogger, getOutputMode } from "../lib/logger/index.ts";
import { handleCommandError } from "../errors/index.ts";
import { GatewayClient, type GatewayDoctorCheck } from "../lib/gateway-client.ts";

interface CheckResult {
  category: string;
  name: string;
  status: "pass" | "fail" | "warn";
  detail: string;
  fix?: string;
}

function mapGatewayCheck(check: GatewayDoctorCheck): CheckResult {
  return {
    category: check.category,
    name: check.name,
    status: check.status,
    detail: check.detail,
    fix: check.fix,
  };
}

export async function runDoctorChecks(): Promise<CheckResult[]> {
  const gateway = new GatewayClient();
  const result = await gateway.doctor();
  if (!result.ok) {
    throw new Error(result.error ?? "Gateway doctor RPC failed");
  }
  return (result.list ?? []).map(mapGatewayCheck);
}

const PASS = "\x1b[32mok\x1b[0m";
const FAIL = "\x1b[31mFAIL\x1b[0m";
const WARN = "\x1b[33mWARN\x1b[0m";

function formatHumanOutput(checks: CheckResult[]): string {
  const lines: string[] = [];
  let currentCategory = "";

  for (const check of checks) {
    if (check.category !== currentCategory) {
      if (currentCategory) lines.push("");
      lines.push(`  ${check.category}`);
      currentCategory = check.category;
    }

    const dots = ".".repeat(Math.max(1, 30 - check.name.length));
    const statusStr = check.status === "pass" ? PASS : check.status === "warn" ? WARN : FAIL;
    const detail = check.detail;
    lines.push(`    ${check.name} ${dots} ${statusStr} (${detail})`);

    if (check.status !== "pass" && check.fix) {
      lines.push(`      \x1b[33mFix: ${check.fix}\x1b[0m`);
    }
  }

  const passed = checks.filter((c) => c.status === "pass").length;
  const failed = checks.filter((c) => c.status === "fail").length;
  const warned = checks.filter((c) => c.status === "warn").length;
  lines.push("");
  lines.push(`  Result: ${passed} passed, ${failed} failed${warned > 0 ? `, ${warned} warnings` : ""}`);

  return lines.join("\n");
}

const doctorCommand = defineCommand({
  meta: {
    name: "doctor",
    description: "Check system prerequisites and vmsan installation health",
  },
  async run() {
    const cmdLog = createCommandLogger("doctor");

    try {
      const checks = await runDoctorChecks();
      const passed = checks.filter((c) => c.status === "pass").length;
      const failed = checks.filter((c) => c.status === "fail").length;

      if (getOutputMode() === "json") {
        cmdLog.set({
          checks: checks.map(({ fix: _fix, ...rest }) => rest),
          summary: { passed, failed, total: checks.length },
        });
      } else {
        consola.log("");
        consola.log("vmsan doctor\n");
        consola.log(formatHumanOutput(checks));
        cmdLog.set({ passed, failed, total: checks.length });
      }

      cmdLog.emit();

      if (failed > 0) {
        process.exitCode = 1;
      }
    } catch (error) {
      handleCommandError(error, cmdLog);
      process.exitCode = 1;
    }
  },
});

export default doctorCommand as CommandDef;

File: hostd/internal/cli/doctor.go
package cli

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/angelorc/vmsan/hostd/internal/gwclient"
	"github.com/angelorc/vmsan/hostd/internal/output"
	"github.com/spf13/cobra"
)

var doctorCmd = &cobra.Command{
	Use:   "doctor",
	Short: "Check system prerequisites and vmsan installation health",
	RunE:  runDoctor,
}

func init() {
	rootCmd.AddCommand(doctorCmd)
}

func runDoctor(_ *cobra.Command, _ []string) error {
	gw, err := gwclient.New()
	if err != nil {
		return fmt.Errorf("gateway connection failed (is the gateway running?): %w", err)
	}
	defer gw.Close()

	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()

	resp, err := gw.Doctor(ctx)
	if err != nil {
		return fmt.Errorf("doctor: %w", err)
	}

	checks := resp.Checks
	passed := 0
	failed := 0
	warned := 0
	for _, c := range checks {
		switch c.Status {
		case "pass":
			passed++
		case "fail":
			failed++
		case "warn":
			warned++
		}
	}

	if jsonOutput {
		type checkJSON struct {
			Category string `json:"category"`
			Name     string `json:"name"`
			Status   string `json:"status"`
			Detail   string `json:"detail"`
		}
		jChecks := make([]checkJSON, len(checks))
		for i, c := range checks {
			jChecks[i] = checkJSON{
				Category: c.Category,
				Name:     c.Name,
				Status:   c.Status,
				Detail:   c.Detail,
			}
		}
		data, _ := json.MarshalIndent(map[string]any{
			"checks":  jChecks,
			"summary": map[string]int{"passed": passed, "failed": failed, "total": len(checks)},
		}, "", "  ")
		fmt.Println(string(data))
		if failed > 0 {
			return fmt.Errorf("%d check(s) failed", failed)
		}
		return nil
	}

	// Human-readable output
	passStr := output.Green + "ok" + output.Reset
	failStr := output.Red + "FAIL" + output.Reset
	warnStr := output.Yellow + "WARN" + output.Reset
	if !output.IsTerminal() {
		passStr = "ok"
		failStr = "FAIL"
		warnStr = "WARN"
	}

	fmt.Println()
	fmt.Println("vmsan doctor")
	fmt.Println()

	currentCategory := ""
	for _, c := range checks {
		if c.Category != currentCategory {
			if currentCategory != "" {
				fmt.Println()
			}
			fmt.Printf("  %s\n", c.Category)
			currentCategory = c.Category
		}

		dots := strings.Repeat(".", max(1, 30-len(c.Name)))
		var statusStr string
		switch c.Status {
		case "pass":
			statusStr = passStr
		case "warn":
			statusStr = warnStr
		default:
			statusStr = failStr
		}
		fmt.Printf("    %s %s %s (%s)\n", c.Name, dots, statusStr, c.Detail)

		if c.Status != "pass" && c.Fix != "" {
			if output.IsTerminal() {
				fmt.Printf("      %sFix: %s%s\n", output.Yellow, c.Fix, output.Reset)
			} else {
				fmt.Printf("      Fix: %s\n", c.Fix)
			}
		}
	}

	fmt.Println()
	summary := fmt.Sprintf("  Result: %d passed, %d failed", passed, failed)
	if warned > 0 {
		summary += fmt.Sprintf(", %d warnings", warned)
	}
	fmt.Println(summary)
	fmt.Println()

	if failed > 0 {
		return fmt.Errorf("%d check(s) failed", failed)
	}
	return nil
}
